### DISCUSSION TITLE
Reservation deletion question

### DISCUSSION THREAD

**Discussion by vals678** [Azure/ipam#223] [Q&A] [answered]
I'm hoping to gain an understanding of the reservation lifecycle. 

- Creating a reservation and assigning the associated tag works as expected. 
- Deletion, from what I've observed:
    - if the reservation hasn't been fulfilled it will update the status to 'cancelledByUser'
    - if the reservation is fulfilled and I want to 'delete' the reservation, nothing happens.
    - Finally, if the VNET associated with a fulfilled reservation is deleted, that IP space is released and can be re-allocated, but the reservation remains in a 'fulfilled' status.

Do the old reservations stay in a perpetually fulfilled status? 
Why would I want to use the reservation process over just requesting the next available VNET endpoint?

**Comment by DCMattyG** [comment, marked as answer]
Hey there @vals678, thanks for the questions. Allow me to answer each one for you...

- **Do the old Reservations stay in a perpetually fulfilled status?**
  - Yes. These serve as a history/audit. Previously these would be deleted, but there were many concerns you would never know whom did what because the historical context was now missing.
- **Why would I want to use the Reservation process over just requesting the next available CIDR address?**
  - The key difference is the Reservation process creates a Reservation and thereby places a "hold" on the CIDR. This stops it from being allocated to other users calling the same API. This is especially important in automation processes where you need to be sure that when you get a CIDR response, that won't change because someone else has consumed that range before the automation completes.

I hope that helps, but certainly if you need any more clarity on the above items please don't hesitate to ask. Very happy to provide as much insight as you need and feedback is always welcome!

**Reply by vals678** [reply]
Thanks @DCMattyG for your reply! This makes sense to me. My one suggestion would be that if the VNET associated via the reservation is deleted/tag removed - that the status is changed to indicate that - either by automation or a call to the IPAM to 'release' a reservation. I think that would help when looking at old reservations. Other than that, works great so far. Thanks!

**Reply by DCMattyG** [reply]
Makes sense @vals678, and that is the exact point of the "fulfilled" status. That is to indicate that the Reservation is now completed (fulfilled), either through cancellation (thereby releasing the CIDR space back into the pool), or via Virtual Network association (thereby releasing the Reservation, but instead the CIDR space is now consumed by the mapped Virtual Network). In either case, a Reservation with a "fulfilled" status no longer affects IP allocations or future CIDR reservation requests.

**Reply by omerzubair** [reply]
in our case @DCMattyG , when we decommission a vnet we do not need historical data we just want the reservation to go back into IP pool. and next time the same range is requested it would pick up the decommissioned space? or perhaps when we are running out of block space then it would pick up the interleaved decommissioned range in the last?
Call to the IPAM to 'release' a reservation maybe?

**Reply by DCMattyG** [reply]
The important this here @omerzubair is that as soon as the reservation has been settled, either by becoming fulfilled or cancelled, it no longer affects future Reservation requests at all.

If you decommission a Virtual Network from Azure (e.g. delete it), you will also need to remove the association from the Block it was associated to in the Azure IPAM tool. Since the tool sits outside of Azure, it's not aware per-se that the Virtual Network was removed, so for safety it marks it as "missing". We do this because we don't want to accidentally remove a Virtual Network association if for some reason an API call to Azure fails and the Virtual Network appears to me missing temporarily.

TL:DR - Settled Reservations (either from becoming fulfilled or cancelled) don't affect future Reservation requests and simply remain for auditing purposes in the future. The only reason you would need to specifically "release" it is if it was never either fulfilled or cancelled in the first place.
